File: tree/python/leetcode105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# Given preorder and inorder traversal of a tree, construct the binary tree.

# Note:
# You may assume that duplicates do not exist in the tree.

# For example, given

# preorder = [3,9,20,15,7]
# inorder = [9,3,15,20,7]
# Return the following binary tree:

#     3
#    / \
#   9  20
#     /  \
#    15   7

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if not preorder or not inorder or len(preorder) == 0 or len(inorder) == 0: return None
        else:
            index = inorder.index(preorder[0])
            root = TreeNode(preorder[0])
            root.left = self.buildTree(preorder[1:index+1], inorder[:index])
            root.right = self.buildTree(preorder[index+1:], inorder[index+1:])
            return root

# Time: O(n)
# Space: O(n)
# Difficulty: medium

File: tree/java/leetcode105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.java
//Given preorder and inorder traversal of a tree, construct the binary tree.
//
//Note:
//You may assume that duplicates do not exist in the tree.
//
//For example, given
//
//preorder = [3,9,20,15,7]
//inorder = [9,3,15,20,7]
//Return the following binary tree:
//
//    3
//   / \
//  9  20
//    /  \
//   15   7

package wenxinjie1;

import java.util.Arrays;

public class leetcode105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal {
    public TreeNode buildTree(int[] preorder, int[] inorder) { 
        
        if (preorder.length == 0 || inorder.length == 0) return null;
        else{
            int index = 0 ;
            for (int i = 0 ; i < inorder.length;i++){
                if (inorder[i] == preorder[0]){
                    index = i;
                    break;
                }
            }
            TreeNode root = new TreeNode(preorder[0]);
            root.left = buildTree(Arrays.copyOfRange(preorder, 1, index+1), Arrays.copyOfRange(inorder, 0, index ));
            root.right = buildTree(Arrays.copyOfRange(preorder, index+1, preorder.length), Arrays.copyOfRange(inorder, index+1, inorder.length));
            return root;
        }     
    }
	
}

//Time: O(n)
//Space: O(n)
//Difficulty: medium